tool/planet/cgroup.go

```diff
@@ -0,0 +1,315 @@
+- Planet services and user tasks take scheduling priority over kubernetes pods
+  - System and User tasks always have priority over pods
+  - kubernetes remains responsible for setting pod cgroup settings, and relative priority between pods
+*/
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"text/template"
+
+	systemdDbus "github.com/coreos/go-systemd/dbus"
+	"github.com/davecgh/go-spew/spew"
+	"github.com/godbus/dbus"
+
+	"github.com/containerd/cgroups"
+	"github.com/gravitational/planet/lib/constants"
+	"github.com/gravitational/planet/lib/utils"
+	"github.com/gravitational/trace"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+)
+
+type CgroupConfig struct {
+	// Enabled indicates whether the planet container should apply the cgroup configuration.
+	Enabled bool
+	// Auto indicates whether planet is allowed to regenerate the Cgroup configuration.
+	// If a new version of planet embeds new rules, this allows planet to update the configuration file. If the
+	// config is externally managed or updated by a user, set this to false so planet doesn't overwrite the settings.
+	Auto bool
+
+	// KubeReservedCPU is the amount of CPU to reserve within kubernetes for kubelet + docker (in millicores)
+	KubeReservedCPU int
+	// KubeSystemCPU is the amount of CPU to reserve within kubernetes for system services (in millicores)
+	KubeSystemCPU int
+
+	// Cgroups is a list of cgroup configurations to apply within the planet container
+	Cgroups []CgroupEntry
+}
+
+type CgroupEntry struct {
+	specs.LinuxResources
+
+	// Path is the cgroup hierarchy path this setting applies to
+	Path []string `json:"path"`
+}
+
+// upsertCgroups reads/updates the cgroup configuration and applies it to the system
+func upsertCgroups(isMaster bool) error {
+	log := logrus.WithField(trace.Component, "cgroup")
+	log.WithField("master", isMaster).Info("Upsert cgroup configuration")
+
+	configPath := path.Join(StateDir, "planet-cgroups.conf")
+	config, err := readCgroupConfig(configPath)
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+	if trace.IsNotFound(err) {
+		config = defaultCgroupConfig(runtime.NumCPU(), isMaster)
+	}
+	if config.Enabled && config.Auto {
+		// if the configuration is automatically generated, regenerate it each start to pick up any changes
+		// with new planet releases
+		config = defaultCgroupConfig(runtime.NumCPU(), isMaster)
+	}
+
+	log.Info("cgroup configuration: ", spew.Sdump(config))
+
+	if !config.Enabled {
+		return nil
+	}
+
+	// rewrite configuration on disk (it may have been updated with new defaults)
+	err = writeCgroupConfig(configPath, config)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	var errors []error
+
+	// try and set the cgroups
+	for _, entry := range config.Cgroups {
+		if len(entry.Path) == 0 {
+			return trace.BadParameter("cgroup spec with no path set: %v", entry)
```

Comment:
Wouldn't this check defeat the purpose after the configuration has already been written to disk?

Reply:
I don't think so, the configuration written to disk, is just so that it's possible for a user or external tool to edit the configuration for the next start of the node. As long as Enabled and Auto are true, the write to disk only really let's you go in and see the config that is being used. However, if it's user specified configuration, it may still have errors.